### ISSUE TITLE
Tracking deprecations

### ISSUE THREAD

**Issue by xificurk** [nette/tester#434]
One of the popular strategies for marking deprecated code in libraries is to use `@trigger_error('...', E_USER_DEPRECATED)` call, see:
- https://github.com/doctrine/deprecations
- https://github.com/symfony/deprecation-contracts

The main advantage and also its main disadvantage is that this method is by default side-effect free. You need to have custom error handler to collect and track usage of the deprecated code.

For PHPUnit there is `symfony/phpunit-bridge` that has quite an extensive feature set for detecting and tracking the deprecations ( https://symfony.com/doc/current/components/phpunit_bridge.html ).

I would like to add at least some basic support for detecting and tracking these deprecations to Nette Tester. Before I start working on any PRs, I would like to discuss, if this even is something that would be accepted. And if yes, what kind of support should I aim for in the first iteration?

**Comment by dg** [comment]
I never understood the point of it. Can you give a specific example of what problem it solves?

**Comment by xificurk** [comment]
Are you asking about tracking the usage of deprecated code? As a user/consumer of the 3rd party package, I would like my CI with Nette Tester to fail, if I use some deprecated API.

**Comment by dg** [comment]
I am asking about `@trigger_error()`.

**Comment by xificurk** [comment]
The main pro-argument I've seen is what I've mentioned - it's side-effect free by default, i.e. the deprecation tracking is opt-in.

But that's not really a problem I want to discuss here... The bottom line is that this strategy is used by various 3rd party packages and it is quite popular. As a user/consumer of those packages and user of Nette Tester, I would like to have way how to detect those deprecations.

**Comment by dg** [comment]
And what do you mean by „detect those deprecations“? Something like a `$scream` from your second pull-request?

**Comment by xificurk** [comment]
Yes, something like that. What exactly "detect" means is one of the things I wanted to discuss.

Detecting these deprecation warnings requires custom error handler, what happens next is the open question. Should the error be simply unmuted? Should it be only logged and reported as soft error at the end of tester run? Should this be configurable? Should there be some kind of support for ignoring specific deprecations? Does it make sense to have various categories for deprecations from direct / indirect calls along the lines of phpunit symfony bridge?

Currently, it's not even very feasible to implement this reliably outside of Nette Tester, because `Assert::error()` registers own error handler, which does not bubble muted errors up to the previously registered error handler.
